File: frontend/src/api/loanAnalysis.ts
/** Loan Analysis API */
import request from './request'

export interface Company {
  id: string
  name: string
  unified_code?: string
  industry?: string
  region?: string
  reg_tags: Record<string, boolean>
  created_at: string
}

export interface CompanyListResult {
  items: Company[]
  total: number
  page: number
  page_size: number
}

export interface DimensionDef {
  id: string
  code: string
  name: string
  weight: number
  scenario: string
  sort_order: number
}

export interface IndicatorDef {
  id: string
  code: string
  name: string
  fibo_path?: string
  formula?: string
  data_source?: string
  unit?: string
  dimension_id?: string
  scenario: string
  weight?: number
  threshold_warning?: number
  threshold_alert?: number
  threshold_direction: string
}

export interface IndicatorValue {
  id: string
  company_id: string
  indicator_id: string
  value?: number
  value_prev?: number
  change_pct?: number
  alert_level: 'normal' | 'warning' | 'alert'
  data_quality?: string
  calc_date: string
  // Enriched
  indicator_name?: string
  indicator_code?: string
  unit?: string
  fibo_path?: string
  formula?: string
  data_source?: string
  threshold_warning?: number
  threshold_alert?: number
  threshold_direction?: string
  dimension_code?: string
  dimension_name?: string
}

export interface DimensionScore {
  name: string
  score: number
  weight: number
  indicator_count: number
  normal_count: number
  warning_count: number
  alert_count: number
}

export interface ScoreRecord {
  id: string
  company_id: string
  scenario: string
  total_score?: number
  risk_level?: string
  dimension_scores: Record<string, DimensionScore>
  suggestion?: string
  calc_date: string
  created_at: string
}

export interface AlertSummary {
  normal: number
  warning: number
  alert: number
}

export interface CompanyScoreResponse {
  company: Company
  score_record?: ScoreRecord
  alert_summary: AlertSummary
  calc_date: string
}

export interface AlertRecord {
  id: string
  company_id: string
  indicator_id: string
  alert_level: 'warning' | 'alert'
  trigger_value?: number
  threshold_value?: number
  trigger_date: string
  action_suggestion?: string
  status: 'open' | 'resolved'
  created_at: string
  indicator_name?: string
  indicator_code?: string
  unit?: string
}

export interface CalculateResult {
  total_score?: number
  risk_level?: string
  alert_count: number
}

// ─── API calls ────────────────────────────────────────────────────

export const loanAnalysisApi = {
  // Companies
  getCompanies: (params?: { page?: number; page_size?: number; search?: string }): Promise<CompanyListResult> =>
    request.get('/loan-analysis/companies', { params }),

  getCompany: (id: string): Promise<Company> =>
    request.get(`/loan-analysis/companies/${id}`),

  createCompany: (data: Partial<Company>): Promise<Company> =>
    request.post('/loan-analysis/companies', data),

  updateCompany: (id: string, data: Partial<Company>): Promise<Company> =>
    request.patch(`/loan-analysis/companies/${id}`, data),

  // Dimensions & Indicators
  getDimensions: (scenario?: string): Promise<DimensionDef[]> =>
    request.get('/loan-analysis/dimensions', { params: scenario ? { scenario } : {} }),

  getIndicators: (scenario?: string): Promise<IndicatorDef[]> =>
    request.get('/loan-analysis/indicators', { params: scenario ? { scenario } : {} }),

  getIndicator: (id: string): Promise<IndicatorDef> =>
    request.get(`/loan-analysis/indicators/${id}`),

  // Indicator values
  getCompanyIndicators: (
    companyId: string,
    scenario: string,
    calcDate?: string,
  ): Promise<IndicatorValue[]> =>
    request.get(`/loan-analysis/companies/${companyId}/indicators`, {
      params: { scenario, calc_date: calcDate },
    }),

  // Score
  getCompanyScore: (
    companyId: string,
    scenario: string,
    calcDate?: string,
  ): Promise<CompanyScoreResponse> =>
    request.get(`/loan-analysis/companies/${companyId}/score`, {
      params: { scenario, calc_date: calcDate },
    }),

  // Calculate
  calculate: (
    companyId: string,
    scenario: string,
    calcDate: string,
  ): Promise<CalculateResult> =>
    request.post(`/loan-analysis/companies/${companyId}/calculate`, {
      scenario,
      calc_date: calcDate,
    }),

  // Alerts
  getAlerts: (
    companyId: string,
    scenario?: string,
    status?: string,
  ): Promise<AlertRecord[]> =>
    request.get(`/loan-analysis/companies/${companyId}/alerts`, {
      params: { scenario, status },
    }),

  // Indicator History Trend
  getIndicatorHistory: (
    companyId: string,
    indicatorId: string,
    params?: { start_date?: string; end_date?: string; limit?: number },
  ): Promise<IndicatorValue[]> =>
    request.get(`/loan-analysis/companies/${companyId}/indicators/${indicatorId}/history`, { params }),

  // Write indicator values with history (dual-write)
  writeIndicatorValuesWithHistory: (
    companyId: string,
    calcDate: string,
    values: Array<{ indicator_id: string; value?: number; value_prev?: number; data_quality?: string }>,
    source: string = 'manual',
  ): Promise<{ history_count: number; upserted_count: number; batch_id: string }> =>
    request.post(`/loan-analysis/companies/${companyId}/values/history?source=${source}`, {
      calc_date: calcDate,
      values,
    }),
}

File: backend/app/schemas/loan_analysis.py
"""Pydantic schemas for the loan analysis API."""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from uuid import UUID


# ─── Company Schemas ────────────────────────────────────────────────

class CompanyCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=256)
    unified_code: Optional[str] = Field(None, max_length=64)
    short_name: Optional[str] = Field(None, max_length=128)
    industry: Optional[str] = Field(None, max_length=128)
    region: Optional[str] = Field(None, max_length=128)
    legal_person: Optional[str] = Field(None, max_length=64)
    registered_capital: Optional[str] = Field(None, max_length=64)
    reg_tags: Optional[Dict[str, Any]] = Field(default_factory=dict)


class CompanyUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=256)
    unified_code: Optional[str] = Field(None, max_length=64)
    short_name: Optional[str] = Field(None, max_length=128)
    industry: Optional[str] = Field(None, max_length=128)
    region: Optional[str] = Field(None, max_length=128)
    legal_person: Optional[str] = Field(None, max_length=64)
    registered_capital: Optional[str] = Field(None, max_length=64)
    is_active: Optional[bool] = None
    reg_tags: Optional[Dict[str, Any]] = None


class CompanyOut(BaseModel):
    id: UUID
    name: str
    unified_code: Optional[str] = None
    short_name: Optional[str] = None
    industry: Optional[str] = None
    region: Optional[str] = None
    legal_person: Optional[str] = None
    registered_capital: Optional[str] = None
    is_active: Optional[bool] = True
    graph_uri: Optional[str] = None
    reg_tags: Dict[str, Any] = {}
    created_at: datetime

    model_config = {"from_attributes": True}


class CompanyListOut(BaseModel):
    items: List[CompanyOut]
    total: int
    page: int
    page_size: int


# ─── Dimension Schemas ───────────────────────────────────────────────

class DimensionOut(BaseModel):
    id: UUID
    code: str
    name: str
    weight: float
    scenario: str
    sort_order: int

    model_config = {"from_attributes": True}


# ─── Indicator Schemas ───────────────────────────────────────────────

class IndicatorOut(BaseModel):
    id: UUID
    code: str
    name: str
    fibo_path: Optional[str]
    formula: Optional[str]
    data_source: Optional[str]
    unit: Optional[str]
    dimension_id: Optional[UUID]
    scenario: str
    weight: Optional[float]
    threshold_warning: Optional[float]
    threshold_alert: Optional[float]
    threshold_direction: str

    model_config = {"from_attributes": True}


class IndicatorWithDimensionOut(IndicatorOut):
    dimension_name: Optional[str] = None
    dimension_code: Optional[str] = None


# ─── Indicator Value Schemas ─────────────────────────────────────────

class IndicatorValueCreate(BaseModel):
    indicator_id: UUID
    value: Optional[float] = None
    value_prev: Optional[float] = None
    calc_date: date
    data_quality: Optional[str] = Field(None, pattern="^(P0|P1|P2)$")


class IndicatorValueOut(BaseModel):
    id: UUID
    company_id: UUID
    indicator_id: UUID
    value: Optional[float]
    value_prev: Optional[float]
    change_pct: Optional[float]
    alert_level: str
    data_quality: Optional[str]
    calc_date: date
    # Enriched fields from joins
    indicator_name: Optional[str] = None
    indicator_code: Optional[str] = None
    unit: Optional[str] = None
    fibo_path: Optional[str] = None
    formula: Optional[str] = None
    data_source: Optional[str] = None
    threshold_warning: Optional[float] = None
    threshold_alert: Optional[float] = None
    threshold_direction: Optional[str] = None
    dimension_code: Optional[str] = None
    dimension_name: Optional[str] = None

    model_config = {"from_attributes": True}


class IndicatorValueBatchCreate(BaseModel):
    company_id: UUID
    calc_date: date
    values: List[IndicatorValueCreate]


# ─── Score Record Schemas ────────────────────────────────────────────

class DimensionScoreDetail(BaseModel):
    score: float
    weight: float
    indicator_count: int
    normal_count: int
    warning_count: int
    alert_count: int


class ScoreRecordOut(BaseModel):
    id: UUID
    company_id: UUID
    scenario: str
    total_score: Optional[float]
    risk_level: Optional[str]
    dimension_scores: Dict[str, Any]
    suggestion: Optional[str]
    calc_date: date
    created_at: datetime

    model_config = {"from_attributes": True}


class CompanyScoreResponse(BaseModel):
    """Complete score response with company info and alert summary."""
    company: CompanyOut
    score_record: Optional[ScoreRecordOut]
    alert_summary: Dict[str, int]  # {"normal": 20, "warning": 6, "alert": 2}
    calc_date: date


# ─── Alert Record Schemas ────────────────────────────────────────────

class AlertRecordOut(BaseModel):
    id: UUID
    company_id: UUID
    indicator_id: UUID
    alert_level: str
    trigger_value: Optional[float]
    threshold_value: Optional[float]
    trigger_date: date
    action_suggestion: Optional[str]
    status: str
    created_at: datetime
    # Enriched
    indicator_name: Optional[str] = None
    indicator_code: Optional[str] = None
    unit: Optional[str] = None

    model_config = {"from_attributes": True}


# ─── Calculate Request ───────────────────────────────────────────────

class CalculateRequest(BaseModel):
    scenario: str = Field(..., pattern="^(pre_loan|post_loan|scf)$")
    calc_date: date
    values: Optional[List[IndicatorValueCreate]] = None  # If None, re-calculate from existing values


# ─── Generic Response ────────────────────────────────────────────────

class ApiResponse(BaseModel):
    code: int = 0
    message: str = "success"
    data: Optional[Any] = None
